### DISCUSSION TITLE
Link mouseenter and mouseleave - how to exclude the tool

### DISCUSSION THREAD

**Discussion by frnora** [clientIO/joint#2230] [Q&A]
### Introduction

Hello guys,

I have a small issue with links and link-tools that keeps bugging me. I have some interactivity features tied to these things which are quite complex and right now I have to solve them in different handlers (element mousedown, stencil drag, etc.) but at the end of the day it could all be solved in one handler... and it boils down to:

1. When a link has a tool and I hover that tool -> the link's `cell:mouseenter` is fired.
2. Leaving the tool then towards the link (if they are touching), hovering it, triggers no event as the link is in theory still "hovered".

I would like to achieve that when I hover the tool, no `cell:mouseenter` is fired. That I can do with `e.stopPropagation()` (I have a custom tool). But then when I leave the tool towards the link, the link should fire `cell:mouseenter`. But it won't because the paper handles this interaction and the tool is "part of the link (view)" and the propagation was stopped already. So what I am trying to do:
1. When a link has a tool and I hover that tool -> **no** `cell:mouseenter` is fired for the link.
2. Leaving the tool then towards the link (if they are touching), hovering it, triggers `cell:mouseenter`. And the other way round with `mouseleave`.

Is there any way to "disconnect" the tool from the linkview for these events? Or do I have to override `Paper.mouseenter` and somehow exclude the link's view lookup through the tool?

### Does your question relate to JointJS or JointJS+. Select both if applicable.

JointJS

**Comment by kumilingus** [comment]
You might try to override the [simulateRelatedView()](https://github.com/clientIO/joint/blob/master/src/dia/ToolView.mjs#L25-L27) to an empty method.

**Reply by frnora** [reply]
Thanks for the suggestion.

This works for `mouseenter` but `mouseleave` doesn't behave well:

If the link is touching MyTool directly (oh that sounds funny...) and I move the mouse from the link to the tool - too many events are fired: 2 mouseleaves, then 1 mouseenter, then 2 mouseleaves again, 1 mouseenter, then 2 mouseleaves. When not overriding `simulateRelatedView`, I get correctly only 1 and 1 event.

If there is another tool between the link and MyTool, let's say a TargetArrow is partially overlapping them both and I move the mouse from link through the arrow to MyTool - `mouseleave` of the paper executes when from link to arrow (this is ignored, as it is a relatedView) but then moving it further to MyTool and out fires nothing. So the basically `mouseleave` of the link never fires in this case.

**Reply by kumilingus** [reply]
Perhaps you could listen to `mouseover` and `mouseleave` and implement the logic based on the `evt.target` (and `evt.relatedTarget`).

```js
paper.on('link:mouseover', function(linkView, evt) {
    console.log(evt.target, 'mouseover');

    const isTargetLink = !!V(evt.target).findParentByClass('joint-link');
    const isTargetTool = !isTargetLink;
});
paper.on('link:mouseout', function(linkView, evt) {
    console.log(evt.target, 'mouseout');

    const isTargetLink = !!V(evt.target).findParentByClass('joint-link');
    const isTargetTool = !isTargetLink;
});
```

**Reply by frnora** [reply]
After struggling with it for a while trying to resolve through the paper with relatedView, etc., I used the simplest possible solution - dispatching an event on mouseenter and mouseout from the tool and then checking the relatedView in the handler.
So something fairly similar to what you have just proposed.

As always, thanks a lot for the suggestion & help.
